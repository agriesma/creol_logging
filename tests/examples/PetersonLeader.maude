load creol-interpreter .
mod PROGRAM is
  protecting CREOL-SIMULATOR .
  op classes : -> Configuration [ctor] .
  eq classes =
    < "Process" : Class | Inh: noInh, Param: "i", "main", Att: "i" |-> null,
      "main" |-> null, "neighbour" |-> null, "mode" |-> null, "status" |->
      null, "uid1" |-> null, "uid2" |-> null,
      Mtds: < "run" : Method | Param: noVid, Att: "_" |-> null, Code: skip ;
              return ( emp ) >,
            < "init" : Method | Param: noVid, Att: "label:1" |-> null,
              "label:0" |-> null, "_" |-> null,
              Code: assign( ( "mode" @ "Process" ), ( "status" @ "Process" ),
              ( "uid1" @ "Process" ) ; bool(true) :: int(0) :: "i" ) ;
              call( "label:0" ; "main" ; "getNeighbour" ; emp ) ;
              get( "label:0" ; "neighbour" ) ; free( "label:0" ) ;
              call( "label:1" ; "neighbour" ; "send" ; "i" ) ;
              free( "label:1" ) ; return ( emp ) >,
            < "send" : Method | Param: "u", Att: "u" |-> null, "label:2" |->
              null, "label:1" |-> null, "label:0" |-> null, "_" |-> null,
              Code: if "mode" th if "=" ( "uid2" :: null ) th if "=" ( "u" ::
              "uid1" ) th assign( "status" ; int(1) ) el
              assign( "uid2" ; "u" ) ; call( "label:0" ; "neighbour" ;
              "send" ; "u" ) ; free( "label:0" ) fi el if "&&" ( ">" ( "uid2"
              :: "uid1" ) :: ">" ( "uid2" :: "u" ) ) th assign( "uid1",
              "uid2" ; "uid2" :: null ) ; call( "label:1" ; "neighbour" ;
              "send" ; "uid1" ) ; free( "label:1" ) el
              assign( "mode" ; bool(false) ) fi fi el call( "label:2" ;
              "neighbour" ; "send" ; "u" ) ; free( "label:2" ) fi ;
              return ( emp ) >,
      Ocnt: 0 >
    
    < "Main" : Class | Inh: noInh, Param: noVid, Att: "p1" |-> null, "p2" |->
      null, "p3" |-> null, "p4" |-> null, "p5" |-> null,
      Mtds: < "run" : Method | Param: noVid, Att: "_" |-> null, Code: skip ;
              return ( emp ) >,
            < "init" : Method | Param: noVid, Att: "_" |-> null,
              Code: new( "p1" ; "Process" ; int(1) :: "this" ) ;
              new( "p2" ; "Process" ; int(3) :: "this" ) ;
              new( "p3" ; "Process" ; int(5) :: "this" ) ;
              new( "p4" ; "Process" ; int(4) :: "this" ) ;
              new( "p5" ; "Process" ; int(2) :: "this" ) ; return ( emp ) >,
            < "getNeighbour" : Method | Param: noVid, Att: "n" |-> null, "_"
              |-> null, Code: if "=" ( "caller" :: "p1" ) th
              assign( "n" ; "p2" ) el if "=" ( "caller" :: "p2" ) th
              assign( "n" ; "p3" ) el if "=" ( "caller" :: "p3" ) th
              assign( "n" ; "p4" ) el if "=" ( "caller" :: "p4" ) th
              assign( "n" ; "p5" ) el if "=" ( "caller" :: "p5" ) th
              assign( "n" ; "p1" ) el skip fi fi fi fi fi ; return ( "n" ) >,
      Ocnt: 0 > .
endm

