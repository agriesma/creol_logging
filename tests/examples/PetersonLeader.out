==========================================
reduce in PROGRAM : classes .
result Configuration: 
< "Main" : Class | Inh: noInh, Param: noVid, Att: "p1" |-> null, "p2" |-> null,
    "p3" |-> null, "p4" |-> null, "p5" |-> null, Mtds:
  < "getNeighbour" : Method | Param: noVid, Att: "_" |-> null, "n" |-> null,
    Code: if "="("caller" :: "p1") th assign("n" ; "p2") el if "="("caller" ::
    "p2") th assign("n" ; "p3") el if "="("caller" :: "p3") th assign("n" ;
    "p4") el if "="("caller" :: "p4") th assign("n" ; "p5") el if "="("caller"
    :: "p5") th assign("n" ; "p1") el skip fi fi fi fi fi ; return("n") >, <
    "init" : Method | Param: noVid, Att: "_" |-> null, Code: new("p1" ;
    "Process" ; int(1) :: "this") ; new("p2" ; "Process" ; int(3) :: "this") ;
    new("p3" ; "Process" ; int(5) :: "this") ; new("p4" ; "Process" ; int(4) ::
    "this") ; new("p5" ; "Process" ; int(2) :: "this") ; return(emp) >, < "run"
    : Method | Param: noVid, Att: "_" |-> null, Code: skip ; return(emp) >,
  Ocnt: 0 >
 
< "Process" : Class | Inh: noInh, Param: "i", "main", Att: "i" |-> null, "main"
    |-> null, "mode" |-> null, "neighbour" |-> null, "status" |-> null, "uid1"
    |-> null, "uid2" |-> null, Mtds:
  < "init" : Method | Param: noVid, Att: "_" |-> null, "label:0" |-> null,
    "label:1" |-> null, Code: assign(("mode" @ "Process"), ("status" @
    "Process"), ("uid1" @ "Process") ; bool(true) :: int(0) :: "i") ; call(
    "label:0" ; "main" ; "getNeighbour" ; emp) ; get("label:0" ; "neighbour") ;
    free("label:0") ; call("label:1" ; "neighbour" ; "send" ; "i") ; free(
    "label:1") ; return(emp) >, < "run" : Method | Param: noVid, Att: "_" |->
    null, Code: skip ; return(emp) >, < "send" : Method | Param: "u", Att: "_"
    |-> null, "label:0" |-> null, "label:1" |-> null, "label:2" |-> null, "u"
    |-> null, Code: if "mode" th if "="("uid2" :: null) th if "="("u" ::
    "uid1") th assign("status" ; int(1)) el assign("uid2" ; "u") ; call(
    "label:0" ; "neighbour" ; "send" ; "u") ; free("label:0") fi el if "&&"(
    ">"("uid2" :: "uid1") :: ">"("uid2" :: "u")) th assign("uid1", "uid2" ;
    "uid2" :: null) ; call("label:1" ; "neighbour" ; "send" ; "uid1") ; free(
    "label:1") el assign("mode" ; bool(false)) fi fi el call("label:2" ;
    "neighbour" ; "send" ; "u") ; free("label:2") fi ; return(emp) >,
  Ocnt: 0 >

==========================================
rewrite in PROGRAM : main(classes, "Main", emp) .
result State: {
< ob("Main0") : "Main" |
  Att: "p1" |-> ob("Process0"), "p2" |-> ob("Process1"), "p3" |-> ob(
    "Process2"), "p4" |-> ob("Process3"), "p5" |-> ob("Process4"), "this" |->
    ob("Main0"),
  Pr: idle,
  PrQ: noProc,
  Lcnt: 2 >
 
< ob("Process0") : "Process" |
  Att: "i" |-> int(1), "main" |-> ob("Main0"), "mode" |-> bool(false),
    "neighbour" |-> ob("Process1"), "status" |-> int(0), "this" |-> ob(
    "Process0"), "uid1" |-> int(1), "uid2" |-> int(2),
  Pr: idle,
  PrQ: noProc,
  Lcnt: 6 >
 
< ob("Process1") : "Process" |
  Att: "i" |-> int(3), "main" |-> ob("Main0"), "mode" |-> bool(false),
    "neighbour" |-> ob("Process2"), "status" |-> int(0), "this" |-> ob(
    "Process1"), "uid1" |-> int(3), "uid2" |-> int(1),
  Pr: idle,
  PrQ: noProc,
  Lcnt: 6 >
 
< ob("Process2") : "Process" |
  Att: "i" |-> int(5), "main" |-> ob("Main0"), "mode" |-> bool(false),
    "neighbour" |-> ob("Process3"), "status" |-> int(0), "this" |-> ob(
    "Process2"), "uid1" |-> int(5), "uid2" |-> int(3),
  Pr: idle,
  PrQ: noProc,
  Lcnt: 6 >
 
< ob("Process3") : "Process" |
  Att: "i" |-> int(4), "main" |-> ob("Main0"), "mode" |-> bool(true),
    "neighbour" |-> ob("Process4"), "status" |-> int(1), "this" |-> ob(
    "Process3"), "uid1" |-> int(5), "uid2" |-> null,
  Pr: idle,
  PrQ: noProc,
  Lcnt: 6 >
 
< ob("Process4") : "Process" |
  Att: "i" |-> int(2), "main" |-> ob("Main0"), "mode" |-> bool(false),
    "neighbour" |-> ob("Process0"), "status" |-> int(0), "this" |-> ob(
    "Process4"), "uid1" |-> int(2), "uid2" |-> int(4),
  Pr: idle,
  PrQ: noProc,
  Lcnt: 6 >
 
< ob("main") : "" |
  Att: noSubst,
  Pr: idle,
  PrQ: noProc,
  Lcnt: 0 >
 
< "Main" : Class | Inh: noInh, Param: noVid, Att: "p1" |-> null, "p2" |-> null,
    "p3" |-> null, "p4" |-> null, "p5" |-> null, Mtds:
  < "getNeighbour" : Method | Param: noVid, Att: "_" |-> null, "n" |-> null,
    Code: if "="("caller" :: "p1") th assign("n" ; "p2") el if "="("caller" ::
    "p2") th assign("n" ; "p3") el if "="("caller" :: "p3") th assign("n" ;
    "p4") el if "="("caller" :: "p4") th assign("n" ; "p5") el if "="("caller"
    :: "p5") th assign("n" ; "p1") el skip fi fi fi fi fi ; return("n") >, <
    "init" : Method | Param: noVid, Att: "_" |-> null, Code: new("p1" ;
    "Process" ; int(1) :: "this") ; new("p2" ; "Process" ; int(3) :: "this") ;
    new("p3" ; "Process" ; int(5) :: "this") ; new("p4" ; "Process" ; int(4) ::
    "this") ; new("p5" ; "Process" ; int(2) :: "this") ; return(emp) >, < "run"
    : Method | Param: noVid, Att: "_" |-> null, Code: skip ; return(emp) >,
  Ocnt: 1 >
 
< "Process" : Class | Inh: noInh, Param: "i", "main", Att: "i" |-> null, "main"
    |-> null, "mode" |-> null, "neighbour" |-> null, "status" |-> null, "uid1"
    |-> null, "uid2" |-> null, Mtds:
  < "init" : Method | Param: noVid, Att: "_" |-> null, "label:0" |-> null,
    "label:1" |-> null, Code: assign(("mode" @ "Process"), ("status" @
    "Process"), ("uid1" @ "Process") ; bool(true) :: int(0) :: "i") ; call(
    "label:0" ; "main" ; "getNeighbour" ; emp) ; get("label:0" ; "neighbour") ;
    free("label:0") ; call("label:1" ; "neighbour" ; "send" ; "i") ; free(
    "label:1") ; return(emp) >, < "run" : Method | Param: noVid, Att: "_" |->
    null, Code: skip ; return(emp) >, < "send" : Method | Param: "u", Att: "_"
    |-> null, "label:0" |-> null, "label:1" |-> null, "label:2" |-> null, "u"
    |-> null, Code: if "mode" th if "="("uid2" :: null) th if "="("u" ::
    "uid1") th assign("status" ; int(1)) el assign("uid2" ; "u") ; call(
    "label:0" ; "neighbour" ; "send" ; "u") ; free("label:0") fi el if "&&"(
    ">"("uid2" :: "uid1") :: ">"("uid2" :: "u")) th assign("uid1", "uid2" ;
    "uid2" :: null) ; call("label:1" ; "neighbour" ; "send" ; "uid1") ; free(
    "label:1") el assign("mode" ; bool(false)) fi fi el call("label:2" ;
    "neighbour" ; "send" ; "u") ; free("label:2") fi ; return(emp) >,
  Ocnt: 5 >
}
Bye.
